components/file_system/src/rate_limiter.rs

```diff
@@ -57,25 +58,35 @@ impl BytesRecorder {
 /// An instance of `IORateLimiter` should be safely shared between threads.
 #[derive(Debug)]
 pub struct IORateLimiter {
-    refill_bytes: AtomicUsize,
-    recorder: Option<Arc<BytesRecorder>>,
+    refill_bytes: usize,
+    enable_statistics: bool,
+    stats: Arc<IORateLimiterStatistics>,
 }
 
 impl IORateLimiter {
-    pub fn new(refill_bytes: usize, recorder: Option<Arc<BytesRecorder>>) -> IORateLimiter {
+    pub fn new(refill_bytes: usize, enable_statistics: bool) -> IORateLimiter {
         IORateLimiter {
-            refill_bytes: AtomicUsize::new(refill_bytes),
-            recorder,
+            refill_bytes,
+            enable_statistics,
+            stats: Arc::new(IORateLimiterStatistics::new()),
         }
     }
 
+    pub fn statistics(&self) -> Arc<IORateLimiterStatistics> {
+        self.stats.clone()
+    }
+
     /// Request for token for bytes and potentially update statistics. If this
     /// request can not be satisfied, the call is blocked. Granted token can be
     /// less than the requested bytes, but must be greater than zero.
     pub fn request(&self, io_type: IOType, io_op: IOOp, bytes: usize) -> usize {
-        let bytes = std::cmp::min(self.refill_bytes.load(atomic::Ordering::Relaxed), bytes);
-        if let Some(recorder) = &self.recorder {
-            recorder.add(io_type, io_op, bytes);
+        let bytes = if self.refill_bytes > 0 {
+            std::cmp::min(self.refill_bytes, bytes)
+        } else {
+            bytes
+        };
+        if self.enable_statistics {
+            self.stats.add(io_type, io_op, bytes);
```

Comment:
can we make the `BytesRecorder` to be a trait `{ fn add(...); fn fetch(...) }` and give an empty `fn add(...)` implementation for IOsnoop. So we don't need to pass the enable_statistics and later io rate limiter can utilize the stats from IOsnoop too.

Reply:
I appreciate your effort to unite the two, my concern is where do we actually need to ignore their differences. For metrics flusher, it's ok; but for IO rate limiter statistics part, it isn't necessary (one could even argue it's worse because a virtual function cal is unlikely to be elided); for IO rate limiter calibration part, the decision to utilize stats is based on whether it is from a external source (calibrate based on our own estimation is a waste of work), which means even if the interface is unified, a extra toggle is required.

Reply:
Also I think wrapping the whole "utilizing external stats" part into a separate object is slightly better than trying to reply on one shared statistics object: the calibration logic needs to maintain some more states than vanilla stats vec; and at this phrase we are really not sure if there will be any other fallback methods to choose from.

Reply:
Okay, I didn't take a look at the calibration logic, so I was thinking it just needs to get the delta bytes periodical. But for the concern of dynamic dispatch, don't worry, you can utilize the https://docs.rs/enum_dispatch/0.3.5/enum_dispatch/ to convert it into static dispatch.